### ISSUE TITLE
Add Weighted Instruction Tuning (WIT) loss to SFTTrainer

### ISSUE THREAD

**Issue by C-anwoy** [huggingface/trl#6746]
## Feature request

I would like to add [Weighted Instruction Tuning (WIT)](https://huggingface.co/papers/2507.07817) as an opt-in loss for `SFTTrainer`. I am one of the paper's authors. The original research implementation is available at https://github.com/kowndinya-renduchintala/WIT.

## Motivation

Conventional instruction tuning assigns zero loss to prompt tokens and unit loss to response tokens. WIT generalizes this by assigning independent weights `lambda_p` and `lambda_r` to prompt and response token losses:

`L_WIT = -sum(lambda_p * log p(prompt tokens) + lambda_r * log p(response tokens)) / N_nonzero`

where `N_nonzero` is the number of tokens whose corresponding prompt/response weight is non-zero. Thus:

- `(lambda_p, lambda_r) = (0, 1)` recovers conventional completion-only SFT.
- `(lambda_p, lambda_r) = (1, 1)` recovers full-sequence language modeling.

The TACL paper evaluates 525 SFT runs across five model configurations and three datasets, and reports that independently weighting prompt and response tokens improves generalization over conventional SFT across the studied settings.

## Proposed TRL API

Following the existing DFT integration through `SFTConfig.loss_type`:

```python
training_args = SFTConfig(
    loss_type="wit",
    prompt_loss_weight=0.2,
    completion_loss_weight=0.8,
)
```

The feature is opt-in. The proposed defaults for WIT-specific weights are `0.0` and `1.0`, matching conventional completion-only instruction tuning.

## Correctness and integration plan

- Preserve the prompt/completion boundary produced during prompt-completion preprocessing and construct per-token weights.
- Mask tokens whose configured weight is zero with `-100`, so Transformers' accumulated `num_items_in_batch` is exactly the WIT denominator across gradient-accumulation microbatches. When `average_tokens_across_devices=True`, current Transformers also aggregates this denominator across data-parallel ranks.
- Compute the weighted sum of shifted token NLLs and divide by `num_items_in_batch`; when it is unavailable (for direct loss calls), divide by the local count of non-zero-weight tokens.
- Support standard and conversational prompt-completion text datasets, padding-free batches, and packing.
- Reject incompatible modes clearly (language-modeling datasets, VLM datasets, assistant-only loss, formatting functions, Liger kernels, custom loss functions, and context/sequence parallelism) rather than silently changing semantics.
- Add unit tests for the formula, zero-weight masking, collation, preprocessing/packing, gradient-accumulation equivalence, validation errors, training, docs, and the paper index.

I searched existing TRL issues and PRs for "Weighted Instruction Tuning", "WIT", and "prompt loss weight" and did not find an existing implementation.

The implementation is available in #6747.
